### PR TITLE
[SPARK-10354] [MLLIB] fix some apparent memory issues in k-means|| initializaiton

### PR DESCRIPTION
**Pull request by mengxr**
- do not cache first cost RDD
- change following cost RDD cache level to MEMORY_AND_DISK
- remove Vector wrapper to save a object per instance

Further improvements will be addressed in SPARK-10329

cc: @yu-iskw @HuJiayin

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala [hunk 1]
```diff
@@ -394,21 +394,28 @@ class KMeans private (
...
-      }.cache()
+          }
+        }.persist(StorageLevel.MEMORY_AND_DISK)
       val sumCosts = costs
-        .aggregate(Vectors.zeros(runs))(
+        .aggregate(new Array[Double](runs))(
           seqOp = (s, v) => {
             // s += v
-            axpy(1.0, v, s)
+            var r = 0
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala]: not calling BLAS here because `runs == 1` in most cases

**Comment by HuJiayin** [comment]
The fix reduces around 50G RDD based on data size below. The performance is improved. The user needs more than 8G memory to run the kmeans in Spark1.5 based on this data size. The data size:
        Number of cluster: 5
	Sample dimensions: 20
	Number of samples: 1200000000
	Sample per input file: 40000000
	K: 10
	Converge distance: 0.5 
	Max iteration: 10

**Comment by HuJiayin** [comment]
LGTM

**Comment by mengxr** [comment]
@HuJiayin Thanks for testing! Merged into master, branch-1.5, 1.4, and 1.3.

**Pull request closed without merging**